crates/sparrow-types/src/instantiate.rs

```diff
@@ -0,0 +1,125 @@
+mod least_upper_bound;
+mod type_constraints;
+
+use arrow_schema::DataType;
+use hashbrown::HashMap;
+use itertools::{izip, Itertools};
+
+use crate::instantiate::type_constraints::TypeConstraints;
+use crate::signature::Signature;
+use crate::type_class::TypeClass;
+use crate::{DisplayFenlType, Error, TypeVariable};
+
+/// The types inferred for a function call.
+#[derive(Debug)]
+pub struct Types {
+    /// The type of each argument to the function call.
+    pub arguments: Vec<DataType>,
+    /// The result of the function call.
+    pub result: DataType,
+}
+
+impl std::fmt::Display for Types {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "({}) -> {}",
+            self.arguments.iter().map(|t| t.display()).format(", "),
+            self.result.display()
+        )
+    }
+}
+
+/// Instantiate arguments to a (possibly) polymorphic signature.
+pub(super) fn instantiate(
+    signature: &Signature,
+    arguments: Vec<DataType>,
+) -> error_stack::Result<Types, Error> {
+    let arguments_len = arguments.len();
+    let parameters = signature.iter_parameters(arguments.len())?;
+
+    // First pass: Collect arguments associated with generic types.
+    let mut type_constraints = TypeConstraints::default();
+
+    for (parameter, argument) in izip!(parameters, arguments) {
+        type_constraints.unify_one(&parameter.ty, argument.into())?;
+    }
+
+    // Solve for each type parameter in the signature.
+    let solutions: HashMap<TypeVariable, DataType> =
+        type_constraints.solve(signature.type_parameters.iter())?;
+
+    // Instantiate the actual types for the signature.
+    let arguments = signature
+        .iter_parameters(arguments_len)?
+        .map(|parameter| {
+            parameter
+                .ty
+                .instantiate(Some(&solutions))
```

Comment:
Currently, `instantiate` is used in two places:

1. When trying to get the "Arrow" type out, if possible. This needs to support both concrete primitives `FenlType::Concrete` and collections like `FenlType::List(FenlType::Concrete(...))`. We can do this by using `instantiate` (which extracts the `DataType` and passing no bindings). If it encounters a `Generic` / Type Variable, it will return `None` or fail.
2. When extracting the result, in which case it shouldn't fail.

I could see a case for making it an error, but the error is different in the two cases. In the first -- it means something went wrong and it's not concrete, like we had expected. In the second, it means something internal went wrong.